### DISCUSSION TITLE
[electron] The sidebar context menu top is different than in VS Code, is it by design or bug?

### DISCUSSION THREAD

**Discussion by kittaakos** [eclipse-theia/theia#12170] [Q&A] [answered]
When I open the context menu from the sidebar, the top of the context menu has different `y` coordinates than the icon.

Theia (TheiaBlueprint Version 1.32.0 (1.32.0.92)):

<img width="294" alt="Screen Shot 2023-02-10 at 16 18 07" src="https://user-images.githubusercontent.com/1405703/218128461-6a8d76eb-6381-42d4-b16e-6b166108555c.png">

VS Code:

<img width="347" alt="Screen Shot 2023-02-10 at 16 18 46" src="https://user-images.githubusercontent.com/1405703/218128527-3e5bab06-bcb8-4bc3-9589-b91ed32214fa.png">

Expected(?) Theia behavior:

![Screen Shot 2023-02-10 at 16 18 07](https://user-images.githubusercontent.com/1405703/218128782-5fd13f00-238c-4a8f-80a7-3b41887160f6.png)

Is it by design, or is it something Theia needs to adjust? Thanks!

**Comment by msujew** [comment, marked as answer]
I believe the sidebar context menu hasn't been touched in a few years. IIRC vscode used to have fewer entries there as well. It might make sense to align to whatever vscode is showing there 👍

**Reply by kittaakos** [reply]
Thanks for the info, Mark.

> IIRC vscode used to have fewer entries there as well. It might make sense to align to whatever vscode is showing there

I was referring to the y coordinates of the context menu and not the items in the context menu. I added a custom sidebar menu, and when I click on it, it pops up below the icon. Is it something Theia wants to change?

Remove `button.height` from here:

https://github.com/eclipse-theia/theia/blob/b8f241309c8bb52269718a38765eab7df4f06897/packages/core/src/browser/shell/sidebar-bottom-menu-widget.tsx#L34

Thank you!

**Reply by msujew** [reply]
I see, yeah I guess that makes sense. We usually want to align to whatever vscode does.

**Reply by kittaakos** [reply]
OK, I know why it's like this. See explanation: 👇 


Behavior with b8f241309:

Electron (normal window): (The context menu `y` is not in line with side-bar menu button `top`.)

<img width="307" alt="Screen Shot 2023-02-14 at 13 55 08" src="https://user-images.githubusercontent.com/1405703/218748561-80916d4a-4daa-4b26-ab5e-a96ffc768508.png">


Electron (maximized window): (It looks OK. electron makes sure the menu fits the screen.)


<img width="333" alt="Screen Shot 2023-02-14 at 13 54 50" src="https://user-images.githubusercontent.com/1405703/218748736-5eaad881-d12e-497e-9b36-9b823c376bb4.png">


Browser: (Looks OK. I guess the menu `y` was picked based on this particular behavior.)

<img width="501" alt="Screen Shot 2023-02-14 at 13 56 11" src="https://user-images.githubusercontent.com/1405703/218749244-f1274c72-fd35-42e1-9ac7-a623f46d7e62.png">


----

Behavior with:

```diff
diff --git a/packages/core/src/browser/shell/sidebar-bottom-menu-widget.tsx b/packages/core/src/browser/shell/sidebar-bottom-menu-widget.tsx
index 351836401..ab17c0218 100644
--- a/packages/core/src/browser/shell/sidebar-bottom-menu-widget.tsx
+++ b/packages/core/src/browser/shell/sidebar-bottom-menu-widget.tsx
@@ -31,7 +31,7 @@ export class SidebarBottomMenuWidget extends SidebarMenuWidget {
             includeAnchorArg: false,
             anchor: {
                 x: button.left + button.width,
-                y: button.top + button.height,
+                y: button.top,
             }
         });
     }
```

Electron (normal window): (This would align the Theia behavior with VS Code.)

<img width="319" alt="Screen Shot 2023-02-14 at 13 58 23" src="https://user-images.githubusercontent.com/1405703/218749414-6754e3d4-abd4-49f6-b9cd-5609e285ca33.png">


Electron (maximized window): (Looks OK. Electron makes the menu fit the screen.)


<img width="332" alt="Screen Shot 2023-02-14 at 13 58 29" src="https://user-images.githubusercontent.com/1405703/218749562-e0a85764-2a05-4120-9887-227ef542f023.png">

Browser: (It looks ugly, and this would be a breaking change in the browser.)

<img width="439" alt="Screen Shot 2023-02-14 at 14 10 19" src="https://user-images.githubusercontent.com/1405703/218749702-191872a1-1f66-4a32-8a1d-6bc1a16444bc.png">

PhosphorJS makes the `y` adjustment [here](https://github.com/phosphorjs/phosphor/blob/9f5e11025b62d2c4a6fb59e2681ae1ed323dcde4/packages/widgets/src/menu.ts#L1480) and subtracts the height of the new menu from the `y` value to make sure it fits the screen.

----

In summary, only not maximized electron windows are affected, and the easy fix breaks the browser. I see a couple of ways to tackle it. if needed: 
 - Downstream adjust the electron code if they will.
 - There can be two service implementations to calculate the anchor from a mouse event.
 - Without any new services, the browser version can calculate `y` in Theia code and use the [`forceY`](https://github.com/phosphorjs/phosphor/blob/9f5e11025b62d2c4a6fb59e2681ae1ed323dcde4/packages/widgets/src/menu.ts#L955) options when [opening](https://github.com/phosphorjs/phosphor/blob/9f5e11025b62d2c4a6fb59e2681ae1ed323dcde4/packages/widgets/src/menu.ts#L437) the menu.
